File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>

int solution(std::vector<std::vector<int>> triangle) {
    std::vector<std::vector<int>> dp(triangle.size(), std::vector<int>(triangle.size(), 0));
    int answer = 0;
    dp[0][0] = triangle[0][0];
    for (int i = 1; i < triangle.size(); i++) {
        for (int j = 0; j <= i; j++) {
            int temp1 = 0, temp2 = 0;
            if (j-1 >= 0) {
                temp1 = dp[i-1][j-1];
            }
            temp2 = dp[i-1][j];
            dp[i][j] = std::max(temp1, temp2) + triangle[i][j];
        }
    }
    for (int i = 0; i < dp[dp.size()-1].size(); i++) {
        answer = std::max(answer, dp[dp.size()-1][i]);
    }
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int[][] sum = new int[triangle.length][];
        for (int i = 0; i < triangle.length; i++) {
            sum[i] = Arrays.copyOf(triangle[i], triangle[i].length);
        }
        for (int i = 1; i < triangle.length; i++) {
            for (int j = 0; j < triangle[i].length; j++) {
                if (j == 0) sum[i][j] = sum[i-1][j]+triangle[i][j];
                else if (j == triangle[i].length-1) sum[i][j] = sum[i-1][j-1]+triangle[i][j];
                else sum[i][j] = Math.max(sum[i-1][j-1]+triangle[i][j], sum[i-1][j]+triangle[i][j]);
                answer = Math.max(answer, sum[i][j]);
            }
        }
        return answer;
    }
}